File: python/kohaku/src/kohaku/spec/action_name.py
"""Rule for resolving the action name a write event invokes (port of TS action-name.ts — the single source of truth)."""

from __future__ import annotations

from .models import JsonObject, UISpec


def resolve_write_action_name(props: JsonObject | None, payload: JsonObject) -> str | None:
    """Resolve the action name for emit=="action.invoke".

    Prefer the target component's props.action (presentForm); otherwise the event payload's action
    (action.button). If neither exists, None (the renderer falls back to forwarding, and the host issues no
    write scope). If action resolution drifts between authorization and execution, writes silently 403, so
    the rule is not duplicated.
    """
    from_props = props.get("action") if props is not None else None
    if isinstance(from_props, str) and len(from_props) > 0:
        return from_props
    from_payload = payload.get("action")
    if isinstance(from_payload, str) and len(from_payload) > 0:
        return from_payload
    return None


def collect_write_actions(spec: UISpec) -> list[str]:
    """Collect the target action names of writes (emit=="action.invoke") from a Spec's events.

    Symmetric to read covering "references the UI reads", this covers "writes the UI declared". Both the
    REST and MCP surfaces consume it so the write-capability issuance rule matches across both profiles.
    """
    by_id = {c.id: c for c in spec.components}
    actions: dict[str, None] = {}
    for e in spec.events:
        if e.emit != "action.invoke":
            continue
        node = by_id.get(e.on.split(".")[0] if e.on else "")
        action = resolve_write_action_name(node.props if node is not None else None, e.payload)
        if action is not None:
            actions[action] = None
    return list(actions.keys())

File: packages/spec-core/src/action-name.ts
import type { JsonObject } from "./schema/json.js";
import type { UISpec } from "./schema/spec.js";

/**
 * The rule for resolving the action name a write event (emit==="action.invoke") invokes (single
 * source of truth). Prefers the target component's props.action (presentForm), falling back to the
 * event payload's action (action.button). If neither is present, undefined (the renderer falls back
 * to forwarding, the host issues no write scope).
 *
 * Both renderer-core (resolveActionName on the execution path) and the host (collectWriteActions for
 * write capability issuance) consume this function. If action resolution for authorization and
 * execution drifted apart, writes would silently 403, so we keep no duplicate of the rule (the
 * "spec-core is the definition site" principle).
 */
export function resolveWriteActionName(
  props: JsonObject | undefined,
  payload: JsonObject,
): string | undefined {
  const fromProps = props?.["action"];
  if (typeof fromProps === "string" && fromProps.length > 0) return fromProps;
  const fromPayload = payload["action"];
  if (typeof fromPayload === "string" && fromPayload.length > 0) return fromPayload;
  return undefined;
}

/**
 * Collects the target action names for writes (emit==="action.invoke") from a Spec's events (single
 * source of truth). Action-name resolution is delegated to resolveWriteActionName (so write-scope
 * issuance for authorization and action resolution on the execution path never diverge). Symmetric to
 * read covering "references the UI reads", this covers "writes the UI declared". Both host-rest
 * (issueCapabilityForSpec on the REST side) and host-mcp-apps (issueCapability on the MCP side)
 * consume this so the write capability issuance rule matches across both profiles.
 */
export function collectWriteActions(spec: UISpec): string[] {
  const byId = new Map(spec.components.map((c) => [c.id, c]));
  const actions = new Set<string>();
  for (const e of spec.events) {
    if (e.emit !== "action.invoke") continue;
    const node = byId.get(e.on.split(".")[0] ?? "");
    const action = resolveWriteActionName(node?.props, e.payload);
    if (action != null) actions.add(action);
  }
  return [...actions];
}
